File: 02-day/main.scala
import scala.io.Source

val movesRaw = Source.fromFile("input.txt").getLines.toList

class Submarine {
  var depth = 0
  var position = 0
  var aim = 0

  def processMoveA(move: String) = {
    val direction = move.split(' ')(0)
    val distance = move.split(' ')(1).toInt

    if (direction == "forward") position += distance 
    else if (direction == "down") depth += distance
    else if (direction == "up") depth -= distance
  }

  def processMoveB(move: String) = {
    val direction = move.split(' ')(0)
    val distance = move.split(' ')(1).toInt

    if (direction == "forward") {
      position += distance
      depth += aim*distance
    } else if (direction == "down") {
      aim += distance
    } else if (direction == "up") {
      aim -= distance
    }
  }

}

val sub1 = new Submarine()
for (move <- movesRaw) sub1.processMoveA(move)
println(sub1.position*sub1.depth)

val sub2 = new Submarine()
for (move <- movesRaw) sub2.processMoveB(move)
println(sub2.position*sub2.depth)

File: 02-day/main.py
with open('input.txt', 'r') as f:
    moves_raw = f.readlines()

class Submarine:
    def __init__(self):
        self.depth = 0
        self.position = 0
        self.aim = 0

    def process_move_a(self, move):
        direction, distance = move.strip().split()

        if direction == 'forward':
            self.position += int(distance)

        elif direction == 'down':
            self.depth += int(distance)

        elif direction == 'up':
            self.depth -= int(distance)

    def process_move_b(self, move):
        direction, distance = move.strip().split()

        if direction == 'forward':
            self.position += int(distance)
            self.depth += self.aim*int(distance)

        elif direction == 'down':
            self.aim += int(distance)

        elif direction == 'up':
            self.aim -= int(distance)

sub = Submarine()
for move in moves_raw:
    sub.process_move_a(move)

res_a = sub.position*sub.depth

print(f"A ::: Result = {res_a}")

sub = Submarine()
for move in moves_raw:
    sub.process_move_b(move)

res_b = sub.position*sub.depth

print(f"B ::: Result = {res_b}")

